File: Moxion.Presentation.Web/ClientApp/src/models/kinematics/VelocityProfile.ts
import Velocity from "../values/Velocity";

class VelocityProfile {
    positiveJerkMaxVelocity: Velocity
    constantAccelerationMaxVelocity: Velocity
    negativeJerkMaxVelocity: Velocity

    constructor(
        positiveJerkMaxVelocity?: Velocity,
        constantAccelerationMaxVelocity?: Velocity,
        negativeJerkMaxVelocity?: Velocity
    ) {
        this.positiveJerkMaxVelocity = positiveJerkMaxVelocity ?? new Velocity()
        this.constantAccelerationMaxVelocity = constantAccelerationMaxVelocity ?? new Velocity()
        this.negativeJerkMaxVelocity = negativeJerkMaxVelocity ?? new Velocity()
    }

    static from(other: VelocityProfile) {
        if (!other) {
            return new VelocityProfile()
        }

        return new VelocityProfile(
            Velocity.from(other.positiveJerkMaxVelocity),
            Velocity.from(other.constantAccelerationMaxVelocity),
            Velocity.from(other.negativeJerkMaxVelocity)
        )
    }
}

export default VelocityProfile

File: Moxion/Domain/Kinematic/VelocityProfile.cs
using Moxion.Abstractions;
using Moxion.Common.Values.Derived;

namespace Moxion.Domain.Kinematic;

public readonly record struct VelocityProfile(
  Velocity PositiveJerkMaxVelocity,
  Velocity ConstantAccelerationMaxVelocity,
  Velocity NegativeJerkMaxVelocity
) : IValueObject<VelocityProfile>;